typing/env.ml

```diff
@@ -2144,6 +2111,41 @@ let read_signature modname filename =
   let ps = read_pers_struct modname filename in
   Lazy.force ps.ps_sig
 
+let add_persistent_structure id env =
+  if not (Ident.persistent id) then invalid_arg "Env.add_persistent_structure";
+  if Ident.name id <> !current_unit then begin
+    { env with
+      modules = IdTbl.add id Persistent env.modules;
+      components = IdTbl.add id Persistent env.components;
+      summary = Env_persistent (env.summary, id);
+    }
+  end else
+    env
+
+let is_identchar_latin1 = function
+  | 'A'..'Z' | 'a'..'z' | '_' | '\192'..'\214' | '\216'..'\246'
+  | '\248'..'\255' | '\'' | '0'..'9' -> true
+  | _ -> false
+
+let unit_name_of_filename fn =
+  match Filename.extension fn with
+  | ".cmi" -> begin
+      let unit =
+        String.capitalize_ascii (Filename.remove_extension fn)
+      in
+      if String.for_all is_identchar_latin1 unit then
+        Some unit
+      else
+        None
+    end
+  | _ -> None
+
+let persistent_structures_of_dir dir =
+  Load_path.Dir.files dir
+  |> List.to_seq
+  |> Seq.filter_map unit_name_of_filename
+  |> String.Set.of_seq
```

Comment:
(One should really have a List.filter_map....)

Reply:
Indeed

Reply:
We do have it - `Misc.Stdlib.List.filter_map`!

Reply:
Indeed, I updated the code